<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Power Simulations for RNA-sequencing • powsimR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Power Simulations for RNA-sequencing">
<meta property="og:description" content="Recent development of very sensitive RNA-seq protocols, such as Smart-seq2 and CEL-seq allows transcriptional
             profiling at single-cell resolution and droplet devices make single cell transcriptomics high-throughput,
             allowing to characterize thousands or even millions of single cells. In powsimR, we have implemented a
             flexible tool to assess power and sample size requirements for differential expression (DE) analysis of single
             cell and bulk RNA-seq experiments. For our read count simulations, we (1) reliably model the mean, dispersion
             and dropout distributions as well as the relationship between those factors from the data. (2) Simulate read
             counts from the empirical mean-variance and dropout relations, while offering flexible choices of the number
             of differentially expressed genes, effect sizes and DE testing method. (3) Finally, we evaluate the power over
             various sample sizes. The number of replicates required to achieve the desired statistical power is mainly
             determined by technical noise and biological variability and both are considerably larger if the biological
             replicates are single cells. powsimR can not only estimate sample sizes necessary to achieve a certain power,
             but also informs about the power to detect DE in a data set at hand. We believe that this type of posterior
             analysis will become more and more important, if results from different studies are to be compared. Often
             enough researchers are left to wonder why there is a lack of overlap in DE-genes across similar experiments.
             PowsimR will allow the researcher to distinguish between actual discrepancies and incongruities due to lack of
             power.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">powsimR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bvieth/powsimR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="powsimr-power-analysis-for-bulk-and-single-cell-rna-seq-experiments" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#powsimr-power-analysis-for-bulk-and-single-cell-rna-seq-experiments" class="anchor"></a>powsimR: Power analysis for bulk and single cell RNA-seq experiments</h1></div>
<p>Please also consult my Github Page of <a href="https://bvieth.github.io/powsimR/">powsimR</a> made with <a href="http://pkgdown.r-lib.org/index.html">pkgdown</a>!</p>
<div id="installation-guide" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-guide" class="anchor"></a>Installation Guide</h2>
<p>For the installation, the R package <code>devtools</code> is needed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"devtools"</span>)
<span class="kw">library</span>(devtools)</code></pre></div>
<p>I recommend to install first the dependencies manually and then powsimR:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ipak &lt;-<span class="st"> </span><span class="cf">function</span>(pkg, <span class="dt">repository =</span> <span class="kw">c</span>(<span class="st">"CRAN"</span>, <span class="st">"Bioconductor"</span>, <span class="st">"github"</span>)) {
    new.pkg &lt;-<span class="st"> </span>pkg[<span class="op">!</span>(pkg <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[, <span class="st">"Package"</span>])]
    <span class="co"># new.pkg &lt;- pkg</span>
    <span class="cf">if</span> (<span class="kw">length</span>(new.pkg)) {
        <span class="cf">if</span> (repository <span class="op">==</span><span class="st"> "CRAN"</span>) {
            <span class="kw">install.packages</span>(new.pkg, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)
        }
        <span class="cf">if</span> (repository <span class="op">==</span><span class="st"> "Bioconductor"</span>) {
            <span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
            <span class="kw">biocLite</span>(new.pkg, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>, <span class="dt">ask =</span> <span class="ot">FALSE</span>)
        }
        <span class="cf">if</span> (repository <span class="op">==</span><span class="st"> "github"</span>) {
            devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(new.pkg, <span class="dt">build_vignettes =</span> <span class="ot">FALSE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)
        }
    }
}

<span class="co"># CRAN PACKAGES</span>
cranpackages &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"bbmle"</span>, <span class="st">"broom"</span>, <span class="st">"cobs"</span>, <span class="st">"cowplot"</span>, <span class="st">"data.table"</span>, <span class="st">"devtools"</span>, 
    <span class="st">"doParallel"</span>, <span class="st">"dplyr"</span>, <span class="st">"drc"</span>, <span class="st">"DrImpute"</span>, <span class="st">"fastICA"</span>, <span class="st">"fitdistrplus"</span>, <span class="st">"foreach"</span>, 
    <span class="st">"gamlss.dist"</span>, <span class="st">"ggExtra"</span>, <span class="st">"ggplot2"</span>, <span class="st">"ggthemes"</span>, <span class="st">"grDevices"</span>, <span class="st">"glmnet"</span>, 
    <span class="st">"grid"</span>, <span class="st">"gtools"</span>, <span class="st">"Hmisc"</span>, <span class="st">"kernlab"</span>, <span class="st">"MASS"</span>, <span class="st">"matrixStats"</span>, <span class="st">"mclust"</span>, <span class="st">"methods"</span>, 
    <span class="st">"minpack.lm"</span>, <span class="st">"moments"</span>, <span class="st">"msir"</span>, <span class="st">"NBPSeq"</span>, <span class="st">"nonnest2"</span>, <span class="st">"parallel"</span>, <span class="st">"penalized"</span>, 
    <span class="st">"plyr"</span>, <span class="st">"pscl"</span>, <span class="st">"reshape2"</span>, <span class="st">"ROCR"</span>, <span class="st">"Rtsne"</span>, <span class="st">"scales"</span>, <span class="st">"Seurat"</span>, <span class="st">"snow"</span>, 
    <span class="st">"stats"</span>, <span class="st">"tibble"</span>, <span class="st">"tidyr"</span>, <span class="st">"VGAM"</span>, <span class="st">"ZIM"</span>)
<span class="kw">ipak</span>(cranpackages, <span class="dt">repository =</span> <span class="st">"CRAN"</span>)

<span class="co"># BIOCONDUCTOR</span>
biocpackages &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"AnnotationDbi"</span>, <span class="st">"baySeq"</span>, <span class="st">"Biobase"</span>, <span class="st">"BiocGenerics"</span>, <span class="st">"BiocParallel"</span>, 
    <span class="st">"DEDS"</span>, <span class="st">"DESeq2"</span>, <span class="st">"EBSeq"</span>, <span class="st">"edgeR"</span>, <span class="st">"IHW"</span>, <span class="st">"iCOBRA"</span>, <span class="st">"limma"</span>, <span class="st">"Linnorm"</span>, 
    <span class="st">"MAST"</span>, <span class="st">"monocle"</span>, <span class="st">"NOISeq"</span>, <span class="st">"qvalue"</span>, <span class="st">"ROTS"</span>, <span class="st">"RUVSeq"</span>, <span class="st">"S4Vectors"</span>, <span class="st">"scater"</span>, 
    <span class="st">"scDD"</span>, <span class="st">"scde"</span>, <span class="st">"scone"</span>, <span class="st">"scran"</span>, <span class="st">"SCnorm"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"SummarizedExperiment"</span>, 
    <span class="st">"zinbwave"</span>)
<span class="kw">ipak</span>(biocpackages, <span class="dt">repository =</span> <span class="st">"Bioconductor"</span>)

<span class="co"># GITHUB</span>
githubpackages &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"nghiavtr/BPSC"</span>, <span class="st">"cz-ye/DECENT"</span>, <span class="st">"mohuangx/SAVER"</span>, <span class="st">"statOmics/zingeR"</span>)
<span class="kw">ipak</span>(githubpackages, <span class="dt">repository =</span> <span class="st">"github"</span>)</code></pre></div>
<p>After installing the dependencies, powsimR can be installed by using devtools as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"bvieth/powsimR"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">FALSE</span>)
<span class="kw">library</span>(<span class="st">"powsimR"</span>)</code></pre></div>
<p>Alternative, you can try to install powsimR and its dependencies directly using devtools:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"bvieth/powsimR"</span>)</code></pre></div>
</div>
<div id="user-guide" class="section level2">
<h2 class="hasAnchor">
<a href="#user-guide" class="anchor"></a>User Guide</h2>
<p>For examples and tips on using the package, please consult the vignette after successful installation by</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">browseVignettes</span>(<span class="st">"powsimR"</span>)</code></pre></div>
<p>Some users have experienced issues installing powsimR due to vignette compilation errors. If that is the case, you can leave out building the vignette and read it on my Github Page of <a href="https://bvieth.github.io/powsimR/articles/powsimR.html">powsimR</a> or download it as a html file <a href="https://github.com/bvieth/powsimR/tree/master/inst/doc/powsimR.html">here</a>.</p>
<div id="dlls-and-ulimit" class="section level3">
<h3 class="hasAnchor">
<a href="#dlls-and-ulimit" class="anchor"></a>DLLs and ulimit</h3>
<p>Note that the error “maximal number of DLLs reached…” might occur due to the loading of many shared objects by Bioconductor packages. Restarting the R session after installing dependencies / powsimR will help. Starting with R version 3.4.0, one can set the environmental variable ‘R_MAX_NUM_DLLS’ to a higher number. See <code>?Startup()</code> for more information. I recommend to increase the maximum number of DLLs that can be loaded to at least 500. The environmental variable R_MAX_NUM_DLLS can be set in R_HOME/etc/Renviron prior to starting R. For that locate the Renviron file and add the following line: R_MAX_NUM_DLLS=xy where xy is the number of DLLs. On my Ubuntu machine, the Renviron file is in /usr/lib/R/etc/ and I can set it to 500.</p>
<p>In addition, the user limits for open files (unix: ulimit) might have to be set to a higher number to accomodate the increase in DLLs. Please check out the help pages for <a href="https://gist.github.com/tombigel/d503800a282fcadbee14b537735d202c">MACs</a> and <a href="https://glassonionblog.wordpress.com/2013/01/27/increase-ulimit-and-file-descriptors-limit/">Linux</a> for guidance.</p>
</div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>Please use the following entry for citing powsimR.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">citation</span>(<span class="st">"powsimR"</span>)</code></pre></div>
<p>powsimR is published in <a href="https://doi.org/10.1093/bioinformatics/btx435">Bioinformatics</a>. A preprint paper is also on <a href="https://doi.org/10.1101/117150">bioRxiv</a>.</p>
</div>
<div id="notes" class="section level2">
<h2 class="hasAnchor">
<a href="#notes" class="anchor"></a>Notes</h2>
<p>Please send bug reports and feature requests by opening a new issue on <a href="https://github.com/bvieth/powsimR/issues">this page</a>.</p>
</div>
<div id="r-session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#r-session-info" class="anchor"></a>R Session Info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(powsimR)
<span class="co">#&gt; Loading required package: gamlss.dist</span>
<span class="co">#&gt; Loading required package: MASS</span>
<span class="co">#&gt; Warning: replacing previous import 'DECENT::lrTest' by 'MAST::lrTest' when</span>
<span class="co">#&gt; loading 'powsimR'</span>
<span class="co">#&gt; Warning: replacing previous import 'parallel::makeCluster' by</span>
<span class="co">#&gt; 'snow::makeCluster' when loading 'powsimR'</span>
<span class="co">#&gt; Warning: replacing previous import 'parallel::stopCluster' by</span>
<span class="co">#&gt; 'snow::stopCluster' when loading 'powsimR'</span>
<span class="co">#&gt; Warning: replacing previous import 'penalized::predict' by 'stats::predict'</span>
<span class="co">#&gt; when loading 'powsimR'</span>
<span class="co">#&gt; Warning: replacing previous import 'zinbwave::glmWeightedF' by</span>
<span class="co">#&gt; 'zingeR::glmWeightedF' when loading 'powsimR'</span>
<span class="kw">sessionInfo</span>()
<span class="co">#&gt; R version 3.5.1 (2018-07-02)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.1 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS: /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] powsimR_1.1.3     gamlss.dist_5.0-6 MASS_7.3-50      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] mixtools_1.1.0              softImpute_1.4             </span>
<span class="co">#&gt;   [3] minpack.lm_1.2-1            pbapply_1.3-4              </span>
<span class="co">#&gt;   [5] haven_1.1.2                 lattice_0.20-35            </span>
<span class="co">#&gt;   [7] fastICA_1.2-1               mgcv_1.8-24                </span>
<span class="co">#&gt;   [9] penalized_0.9-51            blob_1.1.1                 </span>
<span class="co">#&gt;  [11] survival_2.42-6             later_0.7.4                </span>
<span class="co">#&gt;  [13] nloptr_1.0.4                DBI_1.0.0                  </span>
<span class="co">#&gt;  [15] R.utils_2.7.0               SingleCellExperiment_1.2.0 </span>
<span class="co">#&gt;  [17] Linnorm_2.4.0               bindr_0.1.1                </span>
<span class="co">#&gt;  [19] zlibbioc_1.26.0             MatrixModels_0.4-1         </span>
<span class="co">#&gt;  [21] pspline_1.0-18              pcaMethods_1.72.0          </span>
<span class="co">#&gt;  [23] SDMTools_1.1-221            htmlwidgets_1.2            </span>
<span class="co">#&gt;  [25] mvtnorm_1.0-8               hdf5r_1.0.0                </span>
<span class="co">#&gt;  [27] UpSetR_1.3.3                parallel_3.5.1             </span>
<span class="co">#&gt;  [29] scater_1.8.4                irlba_2.3.2                </span>
<span class="co">#&gt;  [31] DEoptimR_1.0-8              lars_1.2                   </span>
<span class="co">#&gt;  [33] Rcpp_0.12.18                KernSmooth_2.23-15         </span>
<span class="co">#&gt;  [35] DT_0.4                      promises_1.0.1             </span>
<span class="co">#&gt;  [37] gdata_2.18.0                DDRTree_0.1.5              </span>
<span class="co">#&gt;  [39] DelayedArray_0.6.5          limma_3.36.3               </span>
<span class="co">#&gt;  [41] vegan_2.5-2                 Hmisc_4.1-1                </span>
<span class="co">#&gt;  [43] ShortRead_1.38.0            apcluster_1.4.7            </span>
<span class="co">#&gt;  [45] RSpectra_0.13-1             fs_1.2.6                   </span>
<span class="co">#&gt;  [47] msir_1.3.1                  mnormt_1.5-5               </span>
<span class="co">#&gt;  [49] digest_0.6.16               png_0.1-7                  </span>
<span class="co">#&gt;  [51] qlcMatrix_0.9.7             cowplot_0.9.3              </span>
<span class="co">#&gt;  [53] glmnet_2.0-16               pkgconfig_2.0.2            </span>
<span class="co">#&gt;  [55] docopt_0.6                  DelayedMatrixStats_1.2.0   </span>
<span class="co">#&gt;  [57] ggbeeswarm_0.6.0            iterators_1.0.10           </span>
<span class="co">#&gt;  [59] minqa_1.2.4                 lavaan_0.6-2               </span>
<span class="co">#&gt;  [61] reticulate_1.10             SummarizedExperiment_1.10.1</span>
<span class="co">#&gt;  [63] spam_2.2-0                  beeswarm_0.2.3             </span>
<span class="co">#&gt;  [65] modeltools_0.2-22           zoo_1.8-3                  </span>
<span class="co">#&gt;  [67] tidyselect_0.2.4            ZIM_1.1.0                  </span>
<span class="co">#&gt;  [69] reshape2_1.4.3              purrr_0.2.5                </span>
<span class="co">#&gt;  [71] kernlab_0.9-27              ica_1.0-2                  </span>
<span class="co">#&gt;  [73] pcaPP_1.9-73                EDASeq_2.14.1              </span>
<span class="co">#&gt;  [75] viridisLite_0.3.0           snow_0.4-2                 </span>
<span class="co">#&gt;  [77] rtracklayer_1.40.6          rlang_0.2.2                </span>
<span class="co">#&gt;  [79] hexbin_1.27.2               manipulateWidget_0.10.0    </span>
<span class="co">#&gt;  [81] glue_1.3.0                  metap_1.0                  </span>
<span class="co">#&gt;  [83] RColorBrewer_1.1-2          registry_0.5               </span>
<span class="co">#&gt;  [85] fpc_2.1-11.1                matrixStats_0.54.0         </span>
<span class="co">#&gt;  [87] stringr_1.3.1               pkgmaker_0.27              </span>
<span class="co">#&gt;  [89] fields_9.6                  DESeq2_1.20.0              </span>
<span class="co">#&gt;  [91] SparseM_1.77                gbRd_0.4-11                </span>
<span class="co">#&gt;  [93] httpuv_1.4.5                class_7.3-14               </span>
<span class="co">#&gt;  [95] BPSC_0.99.2                 RMTstat_0.3                </span>
<span class="co">#&gt;  [97] annotate_1.58.0             webshot_0.5.0              </span>
<span class="co">#&gt;  [99] jsonlite_1.5                XVector_0.20.0             </span>
<span class="co">#&gt; [101] bit_1.1-14                  mime_0.5                   </span>
<span class="co">#&gt; [103] gridExtra_2.3               gplots_3.0.1               </span>
<span class="co">#&gt; [105] Rsamtools_1.32.3            zingeR_0.1.0               </span>
<span class="co">#&gt; [107] stringi_1.2.4               gmodels_2.18.1             </span>
<span class="co">#&gt; [109] gsl_1.9-10.3                bitops_1.0-6               </span>
<span class="co">#&gt; [111] Rdpack_0.9-0                maps_3.3.0                 </span>
<span class="co">#&gt; [113] RSQLite_2.1.1               tidyr_0.8.1                </span>
<span class="co">#&gt; [115] pheatmap_1.0.10             data.table_1.11.4          </span>
<span class="co">#&gt; [117] DEDS_1.54.0                 energy_1.7-5               </span>
<span class="co">#&gt; [119] rstudioapi_0.7              GenomicAlignments_1.16.0   </span>
<span class="co">#&gt; [121] nlme_3.1-137                qvalue_2.12.0              </span>
<span class="co">#&gt; [123] scran_1.8.4                 fastcluster_1.1.25         </span>
<span class="co">#&gt; [125] scone_1.4.0                 locfit_1.5-9.1             </span>
<span class="co">#&gt; [127] miniUI_0.1.1.1              cobs_1.3-3                 </span>
<span class="co">#&gt; [129] R.oo_1.22.0                 prabclus_2.2-6             </span>
<span class="co">#&gt; [131] segmented_0.5-3.0           BiocGenerics_0.26.0        </span>
<span class="co">#&gt; [133] readxl_1.1.0                ROTS_1.8.0                 </span>
<span class="co">#&gt; [135] cellranger_1.1.0            commonmark_1.5             </span>
<span class="co">#&gt; [137] munsell_0.5.0               R.methodsS3_1.7.1          </span>
<span class="co">#&gt; [139] moments_0.14                hwriter_1.3.2              </span>
<span class="co">#&gt; [141] caTools_1.17.1.1            codetools_0.2-15           </span>
<span class="co">#&gt; [143] coda_0.19-1                 Biobase_2.40.0             </span>
<span class="co">#&gt; [145] GenomeInfoDb_1.16.0         vipor_0.4.5                </span>
<span class="co">#&gt; [147] lmtest_0.9-36               htmlTable_1.12             </span>
<span class="co">#&gt; [149] rARPACK_0.11-0              lsei_1.2-0                 </span>
<span class="co">#&gt; [151] xtable_1.8-3                SAVER_1.1.0                </span>
<span class="co">#&gt; [153] ROCR_1.0-7                  diptest_0.75-7             </span>
<span class="co">#&gt; [155] formatR_1.5                 lpsymphony_1.8.0           </span>
<span class="co">#&gt; [157] abind_1.4-5                 FNN_1.1.2.1                </span>
<span class="co">#&gt; [159] RANN_2.6                    sparsesvd_0.1-4            </span>
<span class="co">#&gt; [161] CompQuadForm_1.4.3          GenomicRanges_1.32.6       </span>
<span class="co">#&gt; [163] bibtex_0.4.2                rgl_0.99.16                </span>
<span class="co">#&gt; [165] tibble_1.4.2                ggdendro_0.1-20            </span>
<span class="co">#&gt; [167] cluster_2.0.7-1             Seurat_2.3.4               </span>
<span class="co">#&gt; [169] Matrix_1.2-14               prettyunits_1.0.2          </span>
<span class="co">#&gt; [171] shinyBS_0.61                ggridges_0.5.0             </span>
<span class="co">#&gt; [173] NOISeq_2.24.0               shinydashboard_0.7.0       </span>
<span class="co">#&gt; [175] mclust_5.4.1                igraph_1.2.2               </span>
<span class="co">#&gt; [177] slam_0.1-43                 testthat_2.0.0             </span>
<span class="co">#&gt; [179] doSNOW_1.0.16               htmltools_0.3.6            </span>
<span class="co">#&gt; [181] yaml_2.2.0                  GenomicFeatures_1.32.2     </span>
<span class="co">#&gt; [183] XML_3.98-1.16               pkgdown_1.1.0              </span>
<span class="co">#&gt; [185] DrImpute_1.0                foreign_0.8-71             </span>
<span class="co">#&gt; [187] withr_2.1.2                 fitdistrplus_1.0-11        </span>
<span class="co">#&gt; [189] BiocParallel_1.14.2         aroma.light_3.10.0         </span>
<span class="co">#&gt; [191] bit64_0.9-7                 rngtools_1.3.1             </span>
<span class="co">#&gt; [193] doRNG_1.7.1                 foreach_1.4.4              </span>
<span class="co">#&gt; [195] robustbase_0.93-2           outliers_0.14              </span>
<span class="co">#&gt; [197] scde_2.8.0                  Biostrings_2.48.0          </span>
<span class="co">#&gt; [199] combinat_0.0-8              rsvd_0.9                   </span>
<span class="co">#&gt; [201] iCOBRA_1.8.0                memoise_1.1.0              </span>
<span class="co">#&gt; [203] evaluate_0.11               VGAM_1.0-6                 </span>
<span class="co">#&gt; [205] nonnest2_0.5-1              forcats_0.3.0              </span>
<span class="co">#&gt; [207] rio_0.5.10                  geneplotter_1.58.0         </span>
<span class="co">#&gt; [209] permute_0.9-4               curl_3.2                   </span>
<span class="co">#&gt; [211] fdrtool_1.2.15              trimcluster_0.1-2.1        </span>
<span class="co">#&gt; [213] acepack_1.4.1               edgeR_3.22.3               </span>
<span class="co">#&gt; [215] checkmate_1.8.5             desc_1.2.0                 </span>
<span class="co">#&gt; [217] npsurv_0.4-0                tensorA_0.36.1             </span>
<span class="co">#&gt; [219] DECENT_0.99.1               ellipse_0.4.1              </span>
<span class="co">#&gt; [221] ggplot2_3.0.0               rjson_0.2.20               </span>
<span class="co">#&gt; [223] openxlsx_4.1.0              ggrepel_0.8.0              </span>
<span class="co">#&gt; [225] distillery_1.0-4            dtw_1.20-1                 </span>
<span class="co">#&gt; [227] scDD_1.4.0                  rprojroot_1.3-2            </span>
<span class="co">#&gt; [229] stabledist_0.7-1            Lmoments_1.2-3             </span>
<span class="co">#&gt; [231] tools_3.5.1                 sandwich_2.5-0             </span>
<span class="co">#&gt; [233] magrittr_1.5                RCurl_1.95-4.11            </span>
<span class="co">#&gt; [235] proxy_0.4-22                car_3.0-2                  </span>
<span class="co">#&gt; [237] pbivnorm_0.6.0              ape_5.1                    </span>
<span class="co">#&gt; [239] bayesm_3.1-0.1              xml2_1.2.0                 </span>
<span class="co">#&gt; [241] EBSeq_1.20.0                httr_1.3.1                 </span>
<span class="co">#&gt; [243] assertthat_0.2.0            rmarkdown_1.10             </span>
<span class="co">#&gt; [245] boot_1.3-20                 R6_2.2.2                   </span>
<span class="co">#&gt; [247] Rhdf5lib_1.2.1              nnet_7.3-12                </span>
<span class="co">#&gt; [249] progress_1.2.0              tximport_1.8.0             </span>
<span class="co">#&gt; [251] genefilter_1.62.0           gtools_3.8.1               </span>
<span class="co">#&gt; [253] statmod_1.4.30              Rook_1.1-1                 </span>
<span class="co">#&gt; [255] rhdf5_2.24.0                splines_3.5.1              </span>
<span class="co">#&gt; [257] carData_3.0-1               colorspace_1.3-2           </span>
<span class="co">#&gt; [259] amap_0.8-16                 stats4_3.5.1               </span>
<span class="co">#&gt; [261] NBPSeq_0.3.0                compositions_1.40-2        </span>
<span class="co">#&gt; [263] base64enc_0.1-3             baySeq_2.14.0              </span>
<span class="co">#&gt; [265] pillar_1.3.0                HSMMSingleCell_0.114.0     </span>
<span class="co">#&gt; [267] bindrcpp_0.2.2              GenomeInfoDbData_1.1.0     </span>
<span class="co">#&gt; [269] plyr_1.8.4                  extRemes_2.0-9             </span>
<span class="co">#&gt; [271] dotCall64_1.0-0             gtable_0.2.0               </span>
<span class="co">#&gt; [273] zip_1.0.0                   SCnorm_1.2.1               </span>
<span class="co">#&gt; [275] monocle_2.8.0               knitr_1.20                 </span>
<span class="co">#&gt; [277] RcppArmadillo_0.9.100.5.0   latticeExtra_0.6-28        </span>
<span class="co">#&gt; [279] biomaRt_2.36.1              IRanges_2.14.11            </span>
<span class="co">#&gt; [281] ADGofTest_0.3               copula_0.999-18            </span>
<span class="co">#&gt; [283] crosstalk_1.0.0             Cairo_1.5-9                </span>
<span class="co">#&gt; [285] doParallel_1.0.11           pscl_1.5.2                 </span>
<span class="co">#&gt; [287] flexmix_2.3-14              quantreg_5.36              </span>
<span class="co">#&gt; [289] AnnotationDbi_1.42.1        broom_0.5.0                </span>
<span class="co">#&gt; [291] scales_1.0.0                arm_1.10-1                 </span>
<span class="co">#&gt; [293] backports_1.1.2             IHW_1.8.0                  </span>
<span class="co">#&gt; [295] S4Vectors_0.18.3            densityClust_0.3           </span>
<span class="co">#&gt; [297] lme4_1.1-18-1               brew_1.0-6                 </span>
<span class="co">#&gt; [299] hms_0.4.2                   DESeq_1.32.0               </span>
<span class="co">#&gt; [301] Rtsne_0.13                  dplyr_0.7.6                </span>
<span class="co">#&gt; [303] shiny_1.1.0                 grid_3.5.1                 </span>
<span class="co">#&gt; [305] numDeriv_2016.8-1           bbmle_1.0.20               </span>
<span class="co">#&gt; [307] lazyeval_0.2.1              dynamicTreeCut_1.63-1      </span>
<span class="co">#&gt; [309] Formula_1.2-3               tsne_0.1-3                 </span>
<span class="co">#&gt; [311] blockmodeling_0.3.1         crayon_1.3.4               </span>
<span class="co">#&gt; [313] MAST_1.6.1                  RUVSeq_1.14.0              </span>
<span class="co">#&gt; [315] viridis_0.5.1               roxygen2_6.1.0             </span>
<span class="co">#&gt; [317] rpart_4.1-13                zinbwave_1.2.0             </span>
<span class="co">#&gt; [319] compiler_3.5.1</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/bvieth/powsimR">https://​github.com/​bvieth/​powsimR</a>
</li>
<li>Report a bug at <br><a href="https://github.com/bvieth/powsimR/issues">https://​github.com/​bvieth/​powsimR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL</small></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing powsimR</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Beate Vieth <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-8415-1695" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Beate Vieth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
