<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Power Simulations for RNA-sequencing • powsimR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Power Simulations for RNA-sequencing">
<meta property="og:description" content="Recent development of very sensitive RNA-seq protocols, such as Smart-seq2 and CEL-seq allows
            transcriptional profiling at single-cell resolution and droplet devices make single cell
            transcriptomics high-throughput, allowing to characterize thousands or even millions of single
            cells. In powsimR, we have implemented a flexible tool to assess power and sample size requirements
            for differential expression (DE) analysis of single cell and bulk RNA-seq experiments. For our read
            count simulations, we (1) reliably model the mean, dispersion and dropout distributions as well as
            the relationship between those factors from the data. (2) Simulate read counts from the empirical
            mean-variance and dropout relations, while offering flexible choices of the number of differentially
            expressed genes, effect sizes and DE testing method. (3) Finally, we evaluate the power over various
            sample sizes. The number of replicates required to achieve the desired statistical power is mainly
            determined by technical noise and biological variability and both are considerably larger if the
            biological replicates are single cells. powsimR can not only estimate sample sizes necessary to
            achieve a certain power, but also informs about the power to detect DE in a data set at hand. We
            believe that this type of posterior analysis will become more and more important, if results from
            different studies are to be compared. Often enough researchers are left to wonder why there is a
            lack of overlap in DE-genes across similar experiments. PowsimR will allow the researcher to
            distinguish between actual discrepancies and incongruities due to lack of power.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">powsimR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bvieth/powsimR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="powsimr-power-analysis-for-bulk-and-single-cell-rna-seq-experiments" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#powsimr-power-analysis-for-bulk-and-single-cell-rna-seq-experiments" class="anchor"></a><code>powsimR</code> <br> Power analysis for bulk and <br> single cell RNA-seq experiments <img src="articles/powsimR.png" align="right" width="200">
</h1></div>
<p>Please also consult my Github Page of <a href="https://bvieth.github.io/powsimR/">powsimR</a> made with <a href="http://pkgdown.r-lib.org/index.html">pkgdown</a>!</p>
<div id="arrow_double_down-installation-guide" class="section level2">
<h2 class="hasAnchor">
<a href="#arrow_double_down-installation-guide" class="anchor"></a>:arrow_double_down: Installation Guide</h2>
<p>For the installation, the R package <code>devtools</code> is needed.</p>
<div class="sourceCode" id="cb1"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">devtools</span>)</pre></div>
<p>I recommend to install first the dependencies manually and then powsimR. If you plan to use MAGIC for imputation, then please follow their <a href="https://github.com/KrishnaswamyLab/MAGIC">instruction</a> to install the python implementation before installing powsimR.</p>
<div class="sourceCode" id="cb2"><pre class="r"><span class="no">ipak</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">pkg</span>, <span class="no">repository</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CRAN"</span>, <span class="st">"Bioconductor"</span>, <span class="st">"github"</span>)) {
    <span class="no">new.pkg</span> <span class="kw">&lt;-</span> <span class="no">pkg</span>[!(<span class="no">pkg</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>()[, <span class="st">"Package"</span>])]
    <span class="co"># new.pkg &lt;- pkg</span>
    <span class="kw">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">new.pkg</span>)) {
        <span class="kw">if</span> (<span class="no">repository</span> <span class="kw">==</span> <span class="st">"CRAN"</span>) {
            <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="no">new.pkg</span>, <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
        }
        <span class="kw">if</span> (<span class="no">repository</span> <span class="kw">==</span> <span class="st">"Bioconductor"</span>) {
            <span class="kw">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="no">version</span><span class="kw">[[</span><span class="st">"version.string"</span>]], <span class="st">" "</span>)<span class="kw">[[</span><span class="fl">1</span>]][<span class="fl">3</span>] <span class="kw">&gt;</span> <span class="st">"4.0.0"</span>) {
                <span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>)) {
                  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
                }
                <span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="no">new.pkg</span>, <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">ask</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
            }
            <span class="kw">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="no">version</span><span class="kw">[[</span><span class="st">"version.string"</span>]], <span class="st">" "</span>)<span class="kw">[[</span><span class="fl">1</span>]][<span class="fl">3</span>] <span class="kw">&lt;</span> <span class="st">"3.6.0"</span>) {
                <span class="fu"><a href="https://rdrr.io/r/base/stop.html">stop</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span>(<span class="st">"powsimR depends on packages and functions that are only available in R 4.0.0 and higher."</span>))
            }
        }
        <span class="kw">if</span> (<span class="no">repository</span> <span class="kw">==</span> <span class="st">"github"</span>) {
            <span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="no">new.pkg</span>, <span class="kw">build_vignettes</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">force</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
                <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
        }
    }
}

<span class="co"># CRAN PACKAGES</span>
<span class="no">cranpackages</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"broom"</span>, <span class="st">"cobs"</span>, <span class="st">"cowplot"</span>, <span class="st">"data.table"</span>, <span class="st">"doParallel"</span>, <span class="st">"dplyr"</span>,
    <span class="st">"DrImpute"</span>, <span class="st">"fastICA"</span>, <span class="st">"fitdistrplus"</span>, <span class="st">"foreach"</span>, <span class="st">"future"</span>, <span class="st">"gamlss.dist"</span>, <span class="st">"ggplot2"</span>,
    <span class="st">"ggpubr"</span>, <span class="st">"ggstance"</span>, <span class="st">"grDevices"</span>, <span class="st">"grid"</span>, <span class="st">"Hmisc"</span>, <span class="st">"kernlab"</span>, <span class="st">"MASS"</span>, <span class="st">"magrittr"</span>,
    <span class="st">"MBESS"</span>, <span class="st">"Matrix"</span>, <span class="st">"matrixStats"</span>, <span class="st">"mclust"</span>, <span class="st">"methods"</span>, <span class="st">"minpack.lm"</span>, <span class="st">"moments"</span>,
    <span class="st">"msir"</span>, <span class="st">"NBPSeq"</span>, <span class="st">"nonnest2"</span>, <span class="st">"parallel"</span>, <span class="st">"penalized"</span>, <span class="st">"plyr"</span>, <span class="st">"pscl"</span>, <span class="st">"reshape2"</span>,
    <span class="st">"Rmagic"</span>, <span class="st">"rsvd"</span>, <span class="st">"Rtsne"</span>, <span class="st">"scales"</span>, <span class="st">"Seurat"</span>, <span class="st">"snow"</span>, <span class="st">"sctransform"</span>, <span class="st">"stats"</span>,
    <span class="st">"tibble"</span>, <span class="st">"tidyr"</span>, <span class="st">"truncnorm"</span>, <span class="st">"VGAM"</span>, <span class="st">"ZIM"</span>, <span class="st">"zoo"</span>)
<span class="fu">ipak</span>(<span class="no">cranpackages</span>, <span class="kw">repository</span> <span class="kw">=</span> <span class="st">"CRAN"</span>)

<span class="co"># BIOCONDUCTOR</span>
<span class="no">biocpackages</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"bayNorm"</span>, <span class="st">"baySeq"</span>, <span class="st">"BiocGenerics"</span>, <span class="st">"BiocParallel"</span>, <span class="st">"DESeq2"</span>,
    <span class="st">"EBSeq"</span>, <span class="st">"edgeR"</span>, <span class="st">"IHW"</span>, <span class="st">"iCOBRA"</span>, <span class="st">"limma"</span>, <span class="st">"Linnorm"</span>, <span class="st">"MAST"</span>, <span class="st">"monocle"</span>, <span class="st">"NOISeq"</span>,
    <span class="st">"qvalue"</span>, <span class="st">"ROTS"</span>, <span class="st">"RUVSeq"</span>, <span class="st">"S4Vectors"</span>, <span class="st">"scater"</span>, <span class="st">"scDD"</span>, <span class="st">"scde"</span>, <span class="st">"scone"</span>, <span class="st">"scran"</span>,
    <span class="st">"SCnorm"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"SummarizedExperiment"</span>, <span class="st">"zinbwave"</span>)
<span class="fu">ipak</span>(<span class="no">biocpackages</span>, <span class="kw">repository</span> <span class="kw">=</span> <span class="st">"Bioconductor"</span>)

<span class="co"># GITHUB</span>
<span class="no">githubpackages</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cz-ye/DECENT"</span>, <span class="st">"nghiavtr/BPSC"</span>, <span class="st">"mohuangx/SAVER"</span>, <span class="st">"statOmics/zingeR"</span>,
    <span class="st">"Vivianstats/scImpute"</span>)
<span class="fu">ipak</span>(<span class="no">githubpackages</span>, <span class="kw">repository</span> <span class="kw">=</span> <span class="st">"github"</span>)</pre></div>
<p>To check whether all dependencies are installed, you can run the following lines:</p>
<div class="sourceCode" id="cb3"><pre class="r">
<span class="no">powsimRdeps</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">Package</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">cranpackages</span>,
                                      <span class="no">biocpackages</span>,
                                      <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="no">githubpackages</span>, <span class="st">"/"</span>), <span class="st">"[["</span>, <span class="fl">2</span>)),
                          <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="no">F</span>)

<span class="no">ip</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>()[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">3</span>:<span class="fl">4</span>)], <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="no">F</span>)

<span class="no">ip.check</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">powsimRdeps</span>,
                  <span class="kw">Version</span> <span class="kw">=</span> <span class="no">ip</span>[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="no">powsimRdeps</span>$<span class="no">Package</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">ip</span>)),<span class="st">"Version"</span>])

<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">ip.check</span>$<span class="no">Version</span>))  <span class="co"># all should be FALSE</span></pre></div>
<p>After installing the dependencies, powsimR can be installed by using devtools as well.</p>
<div class="sourceCode" id="cb4"><pre class="r"><span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"bvieth/powsimR"</span>, <span class="kw">build_vignettes</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"powsimR"</span>)</pre></div>
<p>Alternative, you can try to install powsimR and its dependencies directly using devtools:</p>
<div class="sourceCode" id="cb5"><pre class="r"><span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"bvieth/powsimR"</span>)</pre></div>
</div>
<div id="user-guide" class="section level2">
<h2 class="hasAnchor">
<a href="#user-guide" class="anchor"></a><span class="emoji" data-emoji="book">📖</span> User Guide</h2>
<p>For examples and tips on using the package, please consult the vignette after successful installation by</p>
<div class="sourceCode" id="cb6"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes</a></span>(<span class="st">"powsimR"</span>)</pre></div>
<p>Some users have experienced issues installing powsimR due to vignette compilation errors. If that is the case, you can leave out building the vignette (by setting build_vignettes to FALSE) and read it on my Github Page of <a href="https://bvieth.github.io/powsimR/articles/powsimR.html">powsimR</a> or download it as a html file <a href="https://github.com/bvieth/powsimR/tree/master/inst/doc/powsimR.html">here</a>.</p>
<div id="dlls-and-ulimit" class="section level3">
<h3 class="hasAnchor">
<a href="#dlls-and-ulimit" class="anchor"></a>DLLs and ulimit</h3>
<p>Note that the error “maximal number of DLLs reached…” might occur due to the loading of many shared objects by Bioconductor packages. Restarting the R session after installing dependencies / powsimR will help. Starting with R version 3.4.0, one can set the environmental variable ‘R_MAX_NUM_DLLS’ to a higher number. See <code>?Startup()</code> for more information. I recommend to increase the maximum number of DLLs that can be loaded to 500. The environmental variable R_MAX_NUM_DLLS can be set in R_HOME/etc/Renviron prior to starting R. For that locate the Renviron file and add the following line: R_MAX_NUM_DLLS=xy where xy is the number of DLLs. On my Ubuntu machine, the Renviron file is in /usr/lib/R/etc/ and I can set it to 500.</p>
<p>In addition, the user limits for open files (unix: ulimit) might have to be set to a higher number to accomodate the increase in DLLs. Please check out the help pages for <a href="https://gist.github.com/tombigel/d503800a282fcadbee14b537735d202c">MACs</a> and <a href="https://glassonionblog.wordpress.com/2013/01/27/increase-ulimit-and-file-descriptors-limit/">Linux</a> for guidance.</p>
</div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a><span class="emoji" data-emoji="scroll">📜</span> Citation</h2>
<p>Please use the following entry for citing powsimR.</p>
<div class="sourceCode" id="cb7"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span>(<span class="st">"powsimR"</span>)</pre></div>
<p>powsimR is published in <a href="https://doi.org/10.1093/bioinformatics/btx435">Bioinformatics</a>. A preprint paper is also on <a href="https://doi.org/10.1101/117150">bioRxiv</a>.</p>
</div>
<div id="incoming_envelope-notes" class="section level2">
<h2 class="hasAnchor">
<a href="#incoming_envelope-notes" class="anchor"></a>:incoming_envelope: Notes</h2>
<p>Please send bug reports and feature requests by opening a new issue on <a href="https://github.com/bvieth/powsimR/issues">this page</a>. I try to keep up to date with new developments / changes of methods implemented in powsimR, but if you encounter run errors while using a certain tool (e.g. for imputation), then I appreciate if you can post this as an <a href="https://github.com/bvieth/powsimR/issues">issue</a>.</p>
</div>
<div id="r-session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#r-session-info" class="anchor"></a><code>R</code> Session Info</h2>
<div class="sourceCode" id="cb8"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">powsimR</span>)
<span class="co">#&gt; Loading required package: gamlss.dist</span>
<span class="co">#&gt; Loading required package: MASS</span>
<span class="co">#&gt; Warning: replacing previous import 'DECENT::lrTest' by 'MAST::lrTest' when</span>
<span class="co">#&gt; loading 'powsimR'</span>
<span class="co">#&gt; Registered S3 methods overwritten by 'lme4':</span>
<span class="co">#&gt;   method                          from</span>
<span class="co">#&gt;   cooks.distance.influence.merMod car </span>
<span class="co">#&gt;   influence.merMod                car </span>
<span class="co">#&gt;   dfbeta.influence.merMod         car </span>
<span class="co">#&gt;   dfbetas.influence.merMod        car</span>
<span class="co">#&gt; Warning: replacing previous import 'penalized::predict' by 'stats::predict' when</span>
<span class="co">#&gt; loading 'powsimR'</span>
<span class="co">#&gt; Warning: replacing previous import 'zinbwave::glmWeightedF' by</span>
<span class="co">#&gt; 'zingeR::glmWeightedF' when loading 'powsimR'</span>
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.1 (2020-06-06)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.4 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] powsimR_1.2.2     gamlss.dist_5.1-6 MASS_7.3-51.6    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] mixtools_1.2.0              softImpute_1.4             </span>
<span class="co">#&gt;   [3] minpack.lm_1.2-1            lattice_0.20-41            </span>
<span class="co">#&gt;   [5] haven_2.3.1                 vctrs_0.3.1                </span>
<span class="co">#&gt;   [7] fastICA_1.2-2               mgcv_1.8-31                </span>
<span class="co">#&gt;   [9] penalized_0.9-51            blob_1.2.1                 </span>
<span class="co">#&gt;  [11] survival_3.2-3              Rmagic_2.0.3               </span>
<span class="co">#&gt;  [13] later_1.1.0.1               nloptr_1.2.2.1             </span>
<span class="co">#&gt;  [15] DBI_1.1.0                   R.utils_2.9.2              </span>
<span class="co">#&gt;  [17] SingleCellExperiment_1.10.1 rappdirs_0.3.1             </span>
<span class="co">#&gt;  [19] Linnorm_2.12.0              dqrng_0.2.1                </span>
<span class="co">#&gt;  [21] jpeg_0.1-8.1                zlibbioc_1.34.0            </span>
<span class="co">#&gt;  [23] MatrixModels_0.4-1          htmlwidgets_1.5.1          </span>
<span class="co">#&gt;  [25] mvtnorm_1.1-1               future_1.17.0              </span>
<span class="co">#&gt;  [27] UpSetR_1.4.0                parallel_4.0.1             </span>
<span class="co">#&gt;  [29] scater_1.16.1               irlba_2.3.3                </span>
<span class="co">#&gt;  [31] DEoptimR_1.0-8              Rcpp_1.0.4.6               </span>
<span class="co">#&gt;  [33] KernSmooth_2.23-17          DT_0.13                    </span>
<span class="co">#&gt;  [35] promises_1.1.1              gdata_2.18.0               </span>
<span class="co">#&gt;  [37] DDRTree_0.1.5               DelayedArray_0.14.0        </span>
<span class="co">#&gt;  [39] limma_3.44.3                vegan_2.5-6                </span>
<span class="co">#&gt;  [41] Hmisc_4.4-0                 ShortRead_1.46.0           </span>
<span class="co">#&gt;  [43] apcluster_1.4.8             RSpectra_0.16-0            </span>
<span class="co">#&gt;  [45] msir_1.3.2                  mnormt_2.0.0               </span>
<span class="co">#&gt;  [47] digest_0.6.25               png_0.1-7                  </span>
<span class="co">#&gt;  [49] qlcMatrix_0.9.7             sctransform_0.2.1          </span>
<span class="co">#&gt;  [51] cowplot_1.0.0               pkgconfig_2.0.3            </span>
<span class="co">#&gt;  [53] docopt_0.6.1                DelayedMatrixStats_1.10.0  </span>
<span class="co">#&gt;  [55] ggbeeswarm_0.6.0            iterators_1.0.12           </span>
<span class="co">#&gt;  [57] minqa_1.2.4                 lavaan_0.6-6               </span>
<span class="co">#&gt;  [59] reticulate_1.16             SummarizedExperiment_1.18.1</span>
<span class="co">#&gt;  [61] spam_2.5-1                  beeswarm_0.2.3             </span>
<span class="co">#&gt;  [63] modeltools_0.2-23           xfun_0.14                  </span>
<span class="co">#&gt;  [65] zoo_1.8-8                   tidyselect_1.1.0           </span>
<span class="co">#&gt;  [67] ZIM_1.1.0                   reshape2_1.4.4             </span>
<span class="co">#&gt;  [69] purrr_0.3.4                 kernlab_0.9-29             </span>
<span class="co">#&gt;  [71] EDASeq_2.22.0               viridisLite_0.3.0          </span>
<span class="co">#&gt;  [73] snow_0.4-3                  rtracklayer_1.48.0         </span>
<span class="co">#&gt;  [75] rlang_0.4.6                 hexbin_1.28.1              </span>
<span class="co">#&gt;  [77] glue_1.4.1                  RColorBrewer_1.1-2         </span>
<span class="co">#&gt;  [79] fpc_2.2-5                   matrixStats_0.56.0         </span>
<span class="co">#&gt;  [81] stringr_1.4.0               fields_10.3                </span>
<span class="co">#&gt;  [83] ggsignif_0.6.0              DESeq2_1.28.1              </span>
<span class="co">#&gt;  [85] SparseM_1.78                httpuv_1.5.4               </span>
<span class="co">#&gt;  [87] class_7.3-17                BPSC_0.99.2                </span>
<span class="co">#&gt;  [89] BiocNeighbors_1.6.0         annotate_1.66.0            </span>
<span class="co">#&gt;  [91] jsonlite_1.6.1              XVector_0.28.0             </span>
<span class="co">#&gt;  [93] tmvnsim_1.0-2               bit_1.1-15.2               </span>
<span class="co">#&gt;  [95] mime_0.9                    gridExtra_2.3              </span>
<span class="co">#&gt;  [97] gplots_3.0.3                Rsamtools_2.4.0            </span>
<span class="co">#&gt;  [99] zingeR_0.1.0                stringi_1.4.6              </span>
<span class="co">#&gt; [101] gmodels_2.18.1              bitops_1.0-6               </span>
<span class="co">#&gt; [103] maps_3.3.0                  RSQLite_2.2.0              </span>
<span class="co">#&gt; [105] tidyr_1.1.0                 pheatmap_1.0.12            </span>
<span class="co">#&gt; [107] data.table_1.12.8           rstudioapi_0.11            </span>
<span class="co">#&gt; [109] GenomicAlignments_1.24.0    nlme_3.1-148               </span>
<span class="co">#&gt; [111] qvalue_2.20.0               scran_1.16.0               </span>
<span class="co">#&gt; [113] fastcluster_1.1.25          locfit_1.5-9.4             </span>
<span class="co">#&gt; [115] scone_1.12.0                listenv_0.8.0              </span>
<span class="co">#&gt; [117] cobs_1.3-4                  R.oo_1.23.0                </span>
<span class="co">#&gt; [119] prabclus_2.3-2              dbplyr_1.4.4               </span>
<span class="co">#&gt; [121] segmented_1.1-0             BiocGenerics_0.34.0        </span>
<span class="co">#&gt; [123] readxl_1.3.1                lifecycle_0.2.0            </span>
<span class="co">#&gt; [125] ROTS_1.16.0                 munsell_0.5.0              </span>
<span class="co">#&gt; [127] cellranger_1.1.0            R.methodsS3_1.8.0          </span>
<span class="co">#&gt; [129] moments_0.14                hwriter_1.3.2              </span>
<span class="co">#&gt; [131] caTools_1.18.0              codetools_0.2-16           </span>
<span class="co">#&gt; [133] coda_0.19-3                 Biobase_2.48.0             </span>
<span class="co">#&gt; [135] GenomeInfoDb_1.24.2         vipor_0.4.5                </span>
<span class="co">#&gt; [137] htmlTable_1.13.3            bayNorm_1.6.0              </span>
<span class="co">#&gt; [139] rARPACK_0.11-0              xtable_1.8-4               </span>
<span class="co">#&gt; [141] SAVER_1.1.2                 ROCR_1.0-11                </span>
<span class="co">#&gt; [143] diptest_0.75-7              formatR_1.7                </span>
<span class="co">#&gt; [145] lpsymphony_1.16.0           abind_1.4-5                </span>
<span class="co">#&gt; [147] FNN_1.1.3                   RANN_2.6.1                 </span>
<span class="co">#&gt; [149] askpass_1.1                 sparsesvd_0.2              </span>
<span class="co">#&gt; [151] CompQuadForm_1.4.3          GenomicRanges_1.40.0       </span>
<span class="co">#&gt; [153] tibble_3.0.1                ggdendro_0.1-20            </span>
<span class="co">#&gt; [155] cluster_2.1.0               future.apply_1.5.0         </span>
<span class="co">#&gt; [157] Matrix_1.2-18               ellipsis_0.3.1             </span>
<span class="co">#&gt; [159] prettyunits_1.1.1           shinyBS_0.61               </span>
<span class="co">#&gt; [161] NOISeq_2.31.0               shinydashboard_0.7.1       </span>
<span class="co">#&gt; [163] mclust_5.4.6                igraph_1.2.5               </span>
<span class="co">#&gt; [165] ggstance_0.3.4              slam_0.1-47                </span>
<span class="co">#&gt; [167] testthat_2.3.2              doSNOW_1.0.18              </span>
<span class="co">#&gt; [169] htmltools_0.5.0             BiocFileCache_1.12.0       </span>
<span class="co">#&gt; [171] yaml_2.2.1                  GenomicFeatures_1.40.0     </span>
<span class="co">#&gt; [173] XML_3.99-0.3                ggpubr_0.3.0               </span>
<span class="co">#&gt; [175] DrImpute_1.0                foreign_0.8-80             </span>
<span class="co">#&gt; [177] fitdistrplus_1.1-1          BiocParallel_1.22.0        </span>
<span class="co">#&gt; [179] aroma.light_3.18.0          bit64_0.9-7                </span>
<span class="co">#&gt; [181] foreach_1.5.0               robustbase_0.93-6          </span>
<span class="co">#&gt; [183] outliers_0.14               Biostrings_2.56.0          </span>
<span class="co">#&gt; [185] combinat_0.0-8              rsvd_1.0.3                 </span>
<span class="co">#&gt; [187] iCOBRA_1.16.0               memoise_1.1.0              </span>
<span class="co">#&gt; [189] evaluate_0.14               VGAM_1.1-3                 </span>
<span class="co">#&gt; [191] nonnest2_0.5-3              forcats_0.5.0              </span>
<span class="co">#&gt; [193] rio_0.5.16                  geneplotter_1.66.0         </span>
<span class="co">#&gt; [195] permute_0.9-5               curl_4.3                   </span>
<span class="co">#&gt; [197] fdrtool_1.2.15              acepack_1.4.1              </span>
<span class="co">#&gt; [199] edgeR_3.30.3                checkmate_2.0.0            </span>
<span class="co">#&gt; [201] truncnorm_1.0-8             DECENT_1.1.0               </span>
<span class="co">#&gt; [203] tensorA_0.36.1              ellipse_0.4.2              </span>
<span class="co">#&gt; [205] ggplot2_3.3.1               openxlsx_4.1.5             </span>
<span class="co">#&gt; [207] rstatix_0.5.0               ggrepel_0.8.2              </span>
<span class="co">#&gt; [209] scDD_1.12.0                 tools_4.0.1                </span>
<span class="co">#&gt; [211] sandwich_2.5-1              magrittr_1.5               </span>
<span class="co">#&gt; [213] RCurl_1.98-1.2              car_3.0-8                  </span>
<span class="co">#&gt; [215] pbivnorm_0.6.0              bayesm_3.1-4               </span>
<span class="co">#&gt; [217] EBSeq_1.28.0                httr_1.4.1                 </span>
<span class="co">#&gt; [219] assertthat_0.2.1            rmarkdown_2.2              </span>
<span class="co">#&gt; [221] boot_1.3-25                 globals_0.12.5             </span>
<span class="co">#&gt; [223] R6_2.4.1                    Rhdf5lib_1.10.0            </span>
<span class="co">#&gt; [225] nnet_7.3-14                 progress_1.2.2             </span>
<span class="co">#&gt; [227] genefilter_1.70.0           gtools_3.8.2               </span>
<span class="co">#&gt; [229] statmod_1.4.34              BiocSingular_1.4.0         </span>
<span class="co">#&gt; [231] rhdf5_2.32.0                splines_4.0.1              </span>
<span class="co">#&gt; [233] carData_3.0-4               colorspace_1.4-1           </span>
<span class="co">#&gt; [235] amap_0.8-18                 generics_0.0.2             </span>
<span class="co">#&gt; [237] stats4_4.0.1                NBPSeq_0.3.0               </span>
<span class="co">#&gt; [239] base64enc_0.1-3             compositions_1.40-5        </span>
<span class="co">#&gt; [241] baySeq_2.22.0               pillar_1.4.4               </span>
<span class="co">#&gt; [243] HSMMSingleCell_1.8.0        GenomeInfoDbData_1.2.3     </span>
<span class="co">#&gt; [245] plyr_1.8.6                  dotCall64_1.0-0            </span>
<span class="co">#&gt; [247] gtable_0.3.0                zip_2.0.4                  </span>
<span class="co">#&gt; [249] SCnorm_1.10.0               monocle_2.16.0             </span>
<span class="co">#&gt; [251] knitr_1.28                  RcppArmadillo_0.9.900.1.0  </span>
<span class="co">#&gt; [253] latticeExtra_0.6-29         biomaRt_2.44.0             </span>
<span class="co">#&gt; [255] IRanges_2.22.2              fastmap_1.0.1              </span>
<span class="co">#&gt; [257] doParallel_1.0.15           pscl_1.5.5                 </span>
<span class="co">#&gt; [259] flexmix_2.3-15              quantreg_5.55              </span>
<span class="co">#&gt; [261] AnnotationDbi_1.50.0        broom_0.5.6                </span>
<span class="co">#&gt; [263] openssl_1.4.1               scales_1.1.1               </span>
<span class="co">#&gt; [265] arm_1.11-1                  backports_1.1.8            </span>
<span class="co">#&gt; [267] plotrix_3.7-8               IHW_1.16.0                 </span>
<span class="co">#&gt; [269] S4Vectors_0.26.1            densityClust_0.3           </span>
<span class="co">#&gt; [271] lme4_1.1-23                 hms_0.5.3                  </span>
<span class="co">#&gt; [273] DESeq_1.39.0                Rtsne_0.15                 </span>
<span class="co">#&gt; [275] dplyr_1.0.0                 shiny_1.4.0.2              </span>
<span class="co">#&gt; [277] grid_4.0.1                  Formula_1.2-3              </span>
<span class="co">#&gt; [279] blockmodeling_0.3.6         crayon_1.3.4               </span>
<span class="co">#&gt; [281] MAST_1.14.0                 RUVSeq_1.22.0              </span>
<span class="co">#&gt; [283] viridis_0.5.1               rpart_4.1-15               </span>
<span class="co">#&gt; [285] zinbwave_1.10.0             compiler_4.0.1</span></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/bvieth/powsimR/">https://​github.com/​bvieth/​powsimR/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/bvieth/powsimR/issues">https://​github.com/​bvieth/​powsimR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL</small></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing powsimR</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Beate Vieth <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Beate Vieth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
