<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Estimate simulation parameters for spike-ins — estimateSpike • powsimR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Estimate simulation parameters for spike-ins — estimateSpike" />
<meta property="og:description" content="This function estimates and returns parameters needed for spike-in count simulations using supplementary code from Kim et al. 2016 (DOI: 10.1038/ncomms9687)." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">powsimR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bvieth/powsimR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate simulation parameters for spike-ins</h1>
    <small class="dont-index">Source: <a href='https://github.com/bvieth/powsimR/blob/master/R/Estimate.R'><code>R/Estimate.R</code></a></small>
    <div class="hidden name"><code>estimateSpike.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function estimates and returns parameters needed for spike-in count simulations using supplementary code from Kim et al. 2016 (DOI: 10.1038/ncomms9687).</p>
    </div>

    <pre class="usage"><span class='fu'>estimateSpike</span>(<span class='no'>spikeData</span>,
<span class='no'>spikeInfo</span>,
<span class='kw'>MeanFragLengths</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>batchData</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>Normalisation</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>'depth'</span>,<span class='st'>'none'</span>),
<span class='kw'>SampleFilter</span> <span class='kw'>=</span> <span class='fl'>3</span>,
<span class='kw'>RNAseq</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"bulk"</span>, <span class='st'>"singlecell"</span>),
<span class='kw'>Protocol</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>'UMI'</span>, <span class='st'>'Read'</span>),
<span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>spikeData</th>
      <td><p>is a count <code>matrix</code>. Rows correspond to spike-ins, columns to samples.
<code>rownames</code> should contain the spike-in names, <code>colnames</code> the sample names.</p></td>
    </tr>
    <tr>
      <th>spikeInfo</th>
      <td><p>is a molecule count <code>matrix</code> of spike-ins. Rows correspond to spike-ins.
The order of rows should be the same as the rows in <code>spikeData</code>.
The rownames should be the same as the rownames of spikeData.
The first column must contain the molecule counts of spike-ins and be named 'SpikeInput'.
The second column can contain the sequence lengths of spike-ins in bases and be named 'Lengths'.</p></td>
    </tr>
    <tr>
      <th>MeanFragLengths</th>
      <td><p>is a numeric vector of the mean fragment length.</p></td>
    </tr>
    <tr>
      <th>batchData</th>
      <td><p>is a <code>data.frame</code> for batch annotation. Rows correspond to samples.
The order of rows should be the same as the columns in <code>spikeData</code>.
The rownames should be the same as the column names of spikeData.
The first column should contain the batch annotation, e.g. 'a' for batch 1, 'b' for batch 2.</p></td>
    </tr>
    <tr>
      <th>Normalisation</th>
      <td><p>is a character value: 'depth' or 'none'. For more information, please consult the details section.</p></td>
    </tr>
    <tr>
      <th>SampleFilter</th>
      <td><p>is a numeric vector indicating the minimal number of MADs (median absolute deviation)
away from the median number of features detected as well as sequencing depth across all samples
so that outlying samples are removed prior to normalisation and parameter estimation.
The default is <code>5</code>, i.e. at least 5 MADs away from the median.
Choose higher values if you want to filter out less samples.
This parameter is particularly important for single cells to ensure reliable parameter estimation.
For more information, please consult <code><a href='https://rdrr.io/pkg/scater/man/isOutlier.html'>isOutlier</a></code>.</p></td>
    </tr>
    <tr>
      <th>RNAseq</th>
      <td><p>is a character value: "bulk" or "singlecell".</p></td>
    </tr>
    <tr>
      <th>Protocol</th>
      <td><p>is a character value defining the type of counts given in <code>countData</code>.
Options are "UMI" (e.g. 10X Genomics, CEL-seq2) or "Read" (e.g. Smart-seq2).</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical value to indicate whether to print function information.
Default is <code>TRUE</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>List object with the following entries:</p>
<dt>normCounts</dt><dd><p>The normalised spike-in read counts <code>data.frame</code>.</p></dd>
<dt>normParams</dt><dd><p>The mean and standard deviation per spike-in using normalised read counts in a <code>data.frame</code>.</p></dd>
<dt>CaptureEfficiency</dt><dd><p>The ad-hoc estimated as well as fitted detection probabilities with confidence interval per spike-in using normalised read counts in a <code>data.frame</code>.</p></dd>
<dt>seqDepth</dt><dd><p>Library size, i.e. total number of reads per library</p></dd>
<dt>sf</dt><dd><p>The estimated library size factors.</p></dd>
<dt>EVGammaThetaEstimates</dt><dd><p>Estimation of the four parameters capturing technical variability,
namely E[\(\gamma\)], Var[\(\gamma\)], E[\(\theta\)] and Var[\(\theta\)].
For more details, please consult supplementary information of Kim et al. 2016 (DOI: 10.1038/ncomms9687).
These estimates are needed for simulating spike-in read counts.</p></dd>
<dt>Input</dt><dd><p>The input spike-in expression <code>matrix</code>, molecule counts <code>data.frame</code> and batch annotation <code>data.frame</code>,
filtered so that only spike-ins with nonzero expression and samples with at least 100 reads are retained.</p></dd>
<dt>Settings</dt><dd><p>Reporting the chosen normalisation framework.</p></dd>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Normalisation methods</p><dl'>
<dt>'depth'</dt><dd><p>applies the depth normalization method as implemented in <code><a href='https://rdrr.io/pkg/scran/man/computeSpikeFactors.html'>computeSpikeFactors</a></code>.</p></dd>
<dt>'none'</dt><dd><p>No normalisation is applied. This approach can be used for prenormalized expression estimates, e.g. TPM/FPKM/RPKM estimated by RSEM, salmon, cufflinks etc.</p></dd>

</dl>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"SmartSeq2_SpikeIns_Read_Counts"</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"SmartSeq2_SpikeInfo"</span>)
<span class='no'>Batches</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span>(<span class='kw'>Batch</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/strsplit.html'>strsplit</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span>(<span class='no'>SmartSeq2_SpikeIns_Read_Counts</span>),
                                    <span class='st'>"_"</span>), <span class='st'>"[["</span>, <span class='fl'>1</span>),
                     <span class='kw'>stringsAsFactors</span> <span class='kw'>=</span> <span class='no'>F</span>,
                     <span class='kw'>row.names</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span>(<span class='no'>SmartSeq2_SpikeIns_Read_Counts</span>))
<span class='co'># estimation</span>
<span class='no'>spike_param</span> <span class='kw'>&lt;-</span> <span class='fu'>estimateSpike</span>(<span class='kw'>spikeData</span> <span class='kw'>=</span> <span class='no'>SmartSeq2_SpikeIns_Read_Counts</span>,
<span class='kw'>spikeInfo</span> <span class='kw'>=</span> <span class='no'>SmartSeq2_SpikeInfo</span>,
<span class='kw'>MeanFragLength</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>batchData</span> <span class='kw'>=</span> <span class='no'>Batches</span>,
<span class='kw'>Normalisation</span> <span class='kw'>=</span> <span class='st'>'depth'</span>)
<span class='co'># plotting</span>
<span class='fu'><a href='plotSpike.html'>plotSpike</a></span>(<span class='kw'>estSpike</span> <span class='kw'>=</span> <span class='no'>spike_param</span>, <span class='kw'>Annot</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Beate Vieth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


