<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Estimate simulation parameters — estimateParam • powsimR</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Estimate simulation parameters — estimateParam" />

<meta property="og:description" content="This function estimates and returns parameters needed for power simulations assuming a negative binomial read count distribution." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">powsimR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bvieth/powsimR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate simulation parameters</h1>
    <small class="dont-index">Source: <a href='https://github.com/bvieth/powsimR/blob/master/R/Param.R'><code>R/Param.R</code></a></small>
    <div class="hidden name"><code>estimateParam.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function estimates and returns parameters needed for power simulations assuming a negative binomial read count distribution.</p>
    
    </div>

    <pre class="usage"><span class='fu'>estimateParam</span>(<span class='no'>countData</span>,
<span class='kw'>batchData</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>spikeData</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>spikeInfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>Lengths</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>MeanFragLengths</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>Distribution</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>'NB'</span>, <span class='st'>'ZINB'</span>),
<span class='kw'>RNAseq</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>'bulk'</span>, <span class='st'>'singlecell'</span>),
<span class='kw'>normalisation</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>'TMM'</span>, <span class='st'>'MR'</span>, <span class='st'>'PosCounts'</span>, <span class='st'>'UQ'</span>, <span class='st'>'scran'</span>, <span class='st'>'Linnorm'</span>,
                  <span class='st'>'SCnorm'</span>, <span class='st'>'RUV'</span>, <span class='st'>'Census'</span>, <span class='st'>'depth'</span>, <span class='st'>'none'</span>),
<span class='kw'>sigma</span> <span class='kw'>=</span> <span class='fl'>1.96</span>,
<span class='kw'>NCores</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>verbose</span><span class='kw'>=</span><span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>countData</th>
      <td><p>is a count <code>matrix</code>.
Rows correspond to genes, columns to samples.
The gene names should be given as rownames without "_" in the names. The samples names should be given as colnames.</p></td>
    </tr>
    <tr>
      <th>batchData</th>
      <td><p>is a <code>data.frame</code> for batch annotation.
Rows correspond to samples. The first column should contain the batches, e.g. 'a', 'b', 'c', etc.</p></td>
    </tr>
    <tr>
      <th>spikeData</th>
      <td><p>is a count <code>matrix</code>.
Rows correspond to spike-ins, columns to samples.
The order of columns should be the same as in the <code>countData</code>.</p></td>
    </tr>
    <tr>
      <th>spikeInfo</th>
      <td><p>is a molecule count <code>matrix</code> of spike-ins.
Rows correspond to spike-ins. The order of rows should be the same as in the <code>spikeData</code>.
The column names should be 'SpikeID' and 'SpikeInput' for molecule counts of spike-ins.</p></td>
    </tr>
    <tr>
      <th>Lengths</th>
      <td><p>is a numeric vector of transcript lengths with the same length and order as the rows in countData.
This variable is only used for internal TPM calculations if Census normalization is specified.</p></td>
    </tr>
    <tr>
      <th>MeanFragLengths</th>
      <td><p>is a numeric vector of mean fragment lengths with the same length as columns in countData.
This variable is only used for internal TPM calculations if Census normalization is specified.</p></td>
    </tr>
    <tr>
      <th>Distribution</th>
      <td><p>is a character value: "NB" for negative binomial or "ZINB" for zero-inflated negative binomial distribution fitting.</p></td>
    </tr>
    <tr>
      <th>RNAseq</th>
      <td><p>is a character value: "bulk" or "singlecell".</p></td>
    </tr>
    <tr>
      <th>normalisation</th>
      <td><p>is a character value: 'TMM', 'MR', 'PosCounts', 'UQ', 'scran', 'Linnorm',
'SCnorm', 'RUV', 'Census', 'depth', 'none'.
For more information, please consult the details section.</p></td>
    </tr>
    <tr>
      <th>sigma</th>
      <td><p>The variability band width for mean-dispersion loess fit defining the prediction interval for read count simulation. Default is 1.96, i.e. 95% interval. For more information see <code><a href='http://www.rdocumentation.org/packages/msir/topics/loess.sd'>loess.sd</a></code>.</p></td>
    </tr>
    <tr>
      <th>NCores</th>
      <td><p>The number of cores for normalisation method SCnorm and Census.
The default <code>NULL</code> means 1 core.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical value to indicate whether to print function information. Default is <code>TRUE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>List object with the following entries</p>
<dt>seqDepth</dt><dd><p>Library size, i.e. total number of reads per library</p></dd>
<dt>means</dt><dd><p>Mean normalized read counts per gene.</p></dd>
<dt>dispersion</dt><dd><p>Dispersion estimate per gene.</p></dd>
<dt>common.dispersion</dt><dd><p>The common dispersion estimate over all genes.</p></dd>
<dt>size</dt><dd><p>Size parameter of the negative binomial distribution, i.e. 1/dispersion.</p></dd>
<dt>p0</dt><dd><p>Probability that the count will be zero per gene.</p></dd>
<dt>meansizefit</dt><dd><p>A loess fit relating log2 mean to log2 size for use in simulating new data (<code><a href='http://www.rdocumentation.org/packages/msir/topics/loess.sd'>loess.sd</a></code>).</p></dd>
<dt>meandispfit</dt><dd><p>A fit relating log2 mean to log2 dispersion used for visualizing mean-variance dependency (<code><a href='http://www.rdocumentation.org/packages/msir/topics/loess.sd'>loess.sd</a></code>).</p></dd>
<dt>p0.cut</dt><dd><p>The knee point of meanp0fit. Log2 mean values above that value have virtually no dropouts.</p></dd>
<dt>grand.dropout</dt><dd><p>The percentage of empty entries in the count matrix.</p></dd>
<dt>sf</dt><dd><p>The estimated library size factor per sample.</p></dd>
<dt>totalS,totalG</dt><dd><p>Number of samples and genes provided.</p></dd>
<dt>estS,estG</dt><dd><p>Number of samples and genes for which parameters can be estimated.</p></dd>
<dt>RNAseq</dt><dd><p>The type of RNAseq: bulk or single cell.</p></dd>
<dt>normFramework</dt><dd><p>The normalisation method used to calculate normalized parameters and library size factors.</p></dd>
<dt>sigma</dt><dd><p>The width of the variability band.</p></dd>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Normalisation methods</p><dl class='dl-horizontal'>
<dt>TMM, UQ</dt><dd><p>employ the edgeR style normalization of weighted trimmed mean of M-values and upperquartile
as implemented in <code><a href='http://www.rdocumentation.org/packages/edgeR/topics/calcNormFactors'>calcNormFactors</a></code>, respectively.</p></dd>
<dt>MR, PosCounts</dt><dd><p>employ the DESeq2 style normalization of median ratio method and a modified geometric mean method
as implemented in <code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/estimateSizeFactors'>estimateSizeFactors</a></code>, respectively.</p></dd>
<dt>scran, SCnorm</dt><dd><p>apply the deconvolution and quantile regression normalization methods developed for sparse RNA-seq data
as implemented in <code><a href='http://www.rdocumentation.org/packages/scran/topics/computeSumFactors'>computeSumFactors</a></code> and <code><a href='http://www.rdocumentation.org/packages/SCnorm/topics/SCnorm'>SCnorm</a></code>, respectively. Spike-ins can also be supplied for both methods via <code>spikeData</code>. Note, however that this means for scran that the normalisation as implemented in <code><a href='http://www.rdocumentation.org/packages/scran/topics/computeSpikeFactors'>computeSpikeFactors</a></code> is also applied to genes (<code>general.use=TRUE</code>).</p></dd>
<dt>Linnorm</dt><dd><p>apply the normalization method for sparse RNA-seq data
as implemented in <code><a href='http://www.rdocumentation.org/packages/Linnorm/topics/Linnorm.Norm'>Linnorm.Norm</a></code>.
For <code>Linnorm</code>, the user can also supply <code>spikeData</code>.</p></dd>
<dt>RUV</dt><dd><p>removes unwanted variation. There are two approaches implemented:
(1) utilizing negative control genes, i.e. spike-ins stored in <code>spikeData</code> (<code><a href='http://www.rdocumentation.org/packages/RUVSeq/topics/RUVg'>RUVg</a></code>).
(2) utilizing replicate samples, i.e. samples for which the covariates of interest are considered constant.
This annotation is stored in <code>batchData</code> (<code><a href='http://www.rdocumentation.org/packages/RUVSeq/topics/RUVs'>RUVs</a></code>).</p></dd>
<dt>Census</dt><dd><p>converts relative measures of TPM/FPKM values into mRNAs per cell (RPC) without the need of spike-in standards.
Census at least needs <code>Lengths</code> for single-end data and preferably <code>MeanFragLengths</code> for paired-end data.
Do not use this algorithm for UMI data!</p></dd>
<dt>depth</dt><dd><p>Sequencing depth normalisation.</p></dd>
<dt>none</dt><dd><p>No normalisation is applied. This approach can be used for prenormalized expression estimates, e.g. cufflinks, RSEM or salmon.</p></dd>
</dl>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'>## using example data set</span>
<span class='fu'>data</span>(<span class='no'>kolodziejczk_cnts</span>)
<span class='no'>kolodziejczk_estparam</span> <span class='kw'>&lt;-</span> <span class='fu'>estimateParam</span>(<span class='kw'>countData</span><span class='kw'>=</span><span class='no'>kolodziejczk_cnts</span>,
                          <span class='kw'>spikeData</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>spikeInfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                          <span class='kw'>Lengths</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>MeanFragLengths</span><span class='kw'>=</span><span class='kw'>NULL</span>,
                          <span class='kw'>Distribution</span><span class='kw'>=</span><span class='st'>'ZINB'</span>,
                          <span class='kw'>RNAseq</span><span class='kw'>=</span><span class='st'>"singlecell"</span>,
                          <span class='kw'>normalisation</span><span class='kw'>=</span><span class='st'>'scran'</span>,
                          <span class='kw'>NCores</span><span class='kw'>=</span><span class='kw'>NULL</span>,
                          <span class='kw'>sigma</span><span class='kw'>=</span><span class='fl'>1.96</span>)
<span class='co'>## simulating single cell RNA-seq experiment</span>
<span class='no'>ngenes</span> <span class='kw'>&lt;-</span> <span class='fl'>10000</span>
<span class='no'>ncells</span> <span class='kw'>&lt;-</span> <span class='fl'>100</span>
<span class='no'>true.means</span> <span class='kw'>&lt;-</span> <span class='fl'>2</span>^<span class='fu'>runif</span>(<span class='no'>ngenes</span>, <span class='fl'>3</span>, <span class='fl'>6</span>)
<span class='no'>true.dispersions</span> <span class='kw'>&lt;-</span> <span class='fl'>3</span> + <span class='fl'>100</span>/<span class='no'>true.means</span>
<span class='no'>sf.values</span> <span class='kw'>&lt;-</span> <span class='fl'>2</span>^<span class='fu'>rnorm</span>(<span class='no'>ncells</span>, <span class='kw'>sd</span><span class='kw'>=</span><span class='fl'>0.5</span>)
<span class='no'>sf.means</span> <span class='kw'>&lt;-</span> <span class='fu'>outer</span>(<span class='no'>true.means</span>, <span class='no'>sf.values</span>, <span class='st'>'*'</span>)
<span class='no'>cnts</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnbinom</span>(<span class='no'>ngenes</span>*<span class='no'>ncells</span>,
                       <span class='kw'>mu</span><span class='kw'>=</span><span class='no'>sf.means</span>, <span class='kw'>size</span><span class='kw'>=</span><span class='fl'>1</span>/<span class='no'>true.dispersions</span>),
               <span class='kw'>ncol</span><span class='kw'>=</span><span class='no'>ncells</span>)
<span class='co'>## estimating negative binomial parameters</span>
<span class='no'>estparam</span> <span class='kw'>&lt;-</span> <span class='fu'>estimateParam</span>(<span class='kw'>countData</span><span class='kw'>=</span><span class='no'>cnts</span>,
                          <span class='kw'>spikeData</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>spikeInfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                          <span class='kw'>Lengths</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>MeanFragLengths</span><span class='kw'>=</span><span class='kw'>NULL</span>,
                          <span class='kw'>Distribution</span><span class='kw'>=</span><span class='st'>'NB'</span>,
                          <span class='kw'>RNAseq</span><span class='kw'>=</span><span class='st'>"singlecell"</span>,
                          <span class='kw'>normalisation</span><span class='kw'>=</span><span class='st'>'scran'</span>,
                          <span class='kw'>NCores</span><span class='kw'>=</span><span class='kw'>NULL</span>,
                          <span class='kw'>sigma</span><span class='kw'>=</span><span class='fl'>1.96</span>)
<span class='fu'><a href='plotParam.html'>plotParam</a></span>(<span class='no'>estparam</span>, <span class='kw'>annot</span><span class='kw'>=</span><span class='no'>F</span>)

<span class='co'>## simulating bulk RNA-seq experiment</span>
<span class='no'>ngenes</span> <span class='kw'>&lt;-</span> <span class='fl'>10000</span>
<span class='no'>nsamples</span> <span class='kw'>&lt;-</span> <span class='fl'>10</span>
<span class='no'>true.means</span> <span class='kw'>&lt;-</span> <span class='fl'>2</span>^<span class='fu'>rnorm</span>(<span class='no'>ngenes</span>, <span class='kw'>mean</span><span class='kw'>=</span><span class='fl'>8</span>, <span class='kw'>sd</span><span class='kw'>=</span><span class='fl'>2</span>)
<span class='no'>true.dispersions</span> <span class='kw'>&lt;-</span> <span class='fl'>3</span>/<span class='no'>true.means</span> + <span class='fl'>0.1</span>
<span class='no'>sf.values</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='no'>nsamples</span>, <span class='kw'>mean</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>sd</span><span class='kw'>=</span><span class='fl'>0.1</span>)
<span class='no'>sf.means</span> <span class='kw'>&lt;-</span> <span class='fu'>outer</span>(<span class='no'>true.means</span>, <span class='no'>sf.values</span>, <span class='st'>'*'</span>)
<span class='no'>cnts</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnbinom</span>(<span class='no'>ngenes</span>*<span class='no'>nsamples</span>,
                       <span class='kw'>mu</span><span class='kw'>=</span><span class='no'>sf.means</span>, <span class='kw'>size</span><span class='kw'>=</span><span class='fl'>1</span>/<span class='no'>true.dispersions</span>),
               <span class='kw'>ncol</span><span class='kw'>=</span><span class='no'>nsamples</span>)
<span class='co'>## estimating negative binomial parameters</span>
<span class='no'>estparam</span> <span class='kw'>&lt;-</span> <span class='fu'>estimateParam</span>(<span class='kw'>countData</span><span class='kw'>=</span><span class='no'>cnts</span>,
                          <span class='kw'>Distribution</span><span class='kw'>=</span><span class='st'>'NB'</span>, <span class='kw'>RNAseq</span><span class='kw'>=</span><span class='st'>"bulk"</span>,
                          <span class='kw'>normalisation</span><span class='kw'>=</span><span class='st'>'MR'</span>, <span class='kw'>sigma</span><span class='kw'>=</span><span class='fl'>1.96</span>)
<span class='fu'><a href='plotParam.html'>plotParam</a></span>(<span class='no'>estparam</span>, <span class='kw'>annot</span><span class='kw'>=</span><span class='no'>F</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Beate Vieth

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Beate Vieth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

