<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Ziegenhain et al. 2017: Gene Expression Matrices — GeneCounts • powsimR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Ziegenhain et al. 2017: Gene Expression Matrices — GeneCounts" />
<meta property="og:description" content="J1 mESC were cultured in ground pluripotency states by 2i/LIF culture as described previously.
The cells were sorted (FACS) and processed in two batches.
Library preparations were performed with different protocols. Furthermore, spike-ins (ERCC) were added.
&quot;CELseq2&quot; : Library preparations were performed with CEL-seq2 on the Fluidigm C1, as described in Hashimshony et al. 2016.
&quot;DropSeq&quot;: Drop-seq protocol by emulsion of single-cells with primer-microbreads, as described in Macosko et al. 2015.
&quot;MARSseq&quot; : MARS-seq protocol from FACS-sorted single cells as described in Jaitin et al. 2014.
&quot;SCRBseq&quot; : SCRB-seq protocol from FACS-sorted single cells as described in Soumillon et al. 2014.
&quot;SmartSeq&quot; : Library preparation was performed with the Fluidigm C1 platform as per manufacturers' protocol.
&quot;SmartSeq2&quot; : Library preparations were performed using the Smart-seq2 protocol, as described in Picelli et al. 2013.

Data Processing:
base-calls were performed with Illumina bcl2fastq.
All cDNA reads were aligned with Star v2.4.0.
SCRB-seq and Drop-seq protocol reads were tagged with its mate-pair UMI and cell barcode; unique reads were obtained.
Smart-seq and Smart-seq2 reads were assigned to genes using Rsubread.
Only cells were considered with at least 1,000,000 reads.
The R Data files are once for Read Counts and once for UMI Counts per Library Preparation Protocol." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">powsimR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bvieth/powsimR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Ziegenhain et al. 2017: Gene Expression Matrices</h1>
    <small class="dont-index">Source: <a href='https://github.com/bvieth/powsimR/blob/master/R/data.R'><code>R/data.R</code></a></small>
    <div class="hidden name"><code>GeneCounts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>J1 mESC were cultured in ground pluripotency states by 2i/LIF culture as described previously.<br />
The cells were sorted (FACS) and processed in two batches.<br />
Library preparations were performed with different protocols. Furthermore, spike-ins (ERCC) were added.<br />
"CELseq2" : Library preparations were performed with CEL-seq2 on the Fluidigm C1, as described in Hashimshony et al. 2016.<br />
"DropSeq": Drop-seq protocol by emulsion of single-cells with primer-microbreads, as described in Macosko et al. 2015.<br />
"MARSseq" : MARS-seq protocol from FACS-sorted single cells as described in Jaitin et al. 2014.<br />
"SCRBseq" : SCRB-seq protocol from FACS-sorted single cells as described in Soumillon et al. 2014.<br />
"SmartSeq" : Library preparation was performed with the Fluidigm C1 platform as per manufacturers' protocol.<br />
"SmartSeq2" : Library preparations were performed using the Smart-seq2 protocol, as described in Picelli et al. 2013.<br />
<br />
Data Processing:<br />
base-calls were performed with Illumina bcl2fastq.<br />
All cDNA reads were aligned with Star v2.4.0.<br />
SCRB-seq and Drop-seq protocol reads were tagged with its mate-pair UMI and cell barcode; unique reads were obtained.<br />
Smart-seq and Smart-seq2 reads were assigned to genes using Rsubread.<br />
Only cells were considered with at least 1,000,000 reads.<br />
The R Data files are once for Read Counts and once for UMI Counts per Library Preparation Protocol.</p>
    </div>

    <pre class="usage"><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>CELseq2_Gene_Read_Counts</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>CELseq2_Gene_UMI_Counts</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>DropSeq_Gene_Read_Counts</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>DropSeq_Gene_UMI_Counts</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>MARSseq_Gene_Read_Counts</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>MARSseq_Gene_UMI_Counts</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>SCRBseq_Gene_Read_Counts</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>SCRBseq_Gene_UMI_Counts</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>SmartSeq_Gene_Read_Counts</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>SmartSeq2_Gene_Read_Counts</span>)</pre>


    <h2 class="hasAnchor" id="format"><a class="anchor" href="#format"></a>Format</h2>

    <p>An object of class <code>"data.frame"</code>, with genes per row and cells per column.</p>
    <h2 class="hasAnchor" id="source"><a class="anchor" href="#source"></a>Source</h2>

    <p><a href='https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75790'>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75790</a>.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Beate Vieth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


