<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Setup options for RNA-seq count simulations — Setup • powsimR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Setup options for RNA-seq count simulations — Setup" />
<meta property="og:description" content="This function generates the settings needed for simulateDE.
Firstly a set of differential expressed gene IDs with
associated log fold changes for a given number of genes, simulations and
fraction of DE genes is generated. There are also a number of options relating
 to count simulations such as downsampling.
 Secondly, the estimated parameters for count simulations of genes (and spikes) as well as samples are added." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">powsimR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bvieth/powsimR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Setup options for RNA-seq count simulations</h1>
    <small class="dont-index">Source: <a href='https://github.com/bvieth/powsimR/blob/master/R/Setup.R'><code>R/Setup.R</code></a></small>
    <div class="hidden name"><code>Setup.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function generates the settings needed for <code><a href='simulateDE.html'>simulateDE</a></code>.
Firstly a set of differential expressed gene IDs with
associated log fold changes for a given number of genes, simulations and
fraction of DE genes is generated. There are also a number of options relating
 to count simulations such as downsampling.
 Secondly, the estimated parameters for count simulations of genes (and spikes) as well as samples are added.</p>
    </div>

    <pre class="usage"><span class='fu'>Setup</span><span class='op'>(</span>ngenes<span class='op'>=</span><span class='cn'>NULL</span>, nsims<span class='op'>=</span><span class='fl'>25</span>,
p.DE <span class='op'>=</span> <span class='fl'>0.1</span>, pLFC <span class='op'>=</span> <span class='fl'>1</span>, p.G <span class='op'>=</span> <span class='fl'>1</span>,
p.B<span class='op'>=</span><span class='cn'>NULL</span>, bLFC<span class='op'>=</span><span class='cn'>NULL</span>, bPattern<span class='op'>=</span><span class='st'>"uncorrelated"</span>,
n1<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>20</span>,<span class='fl'>50</span>,<span class='fl'>100</span><span class='op'>)</span>, n2<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>30</span>,<span class='fl'>60</span>,<span class='fl'>120</span><span class='op'>)</span>,
Thinning <span class='op'>=</span> <span class='cn'>NULL</span>, LibSize<span class='op'>=</span><span class='st'>'equal'</span>,
<span class='va'>estParamRes</span>, estSpikeRes<span class='op'>=</span><span class='cn'>NULL</span>,
DropGenes<span class='op'>=</span><span class='cn'>FALSE</span>,
<span class='va'>setup.seed</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ngenes</th>
      <td><p>is a numeric vector specifying the number of genes to simulate.
Default is <code>NULL</code>, i.e. the number of genes that were deemed expressed after filtering.
See <code><a href='estimateParam.html'>estimateParam</a></code> and <code><a href='plotParam.html'>plotParam</a></code> for more information about the number of genes that were used fo estimation and fitting.</p></td>
    </tr>
    <tr>
      <th>nsims</th>
      <td><p>Number of simulations to run. Default is 25.</p></td>
    </tr>
    <tr>
      <th>p.DE</th>
      <td><p>Numeric vector between 0 and 1 representing
the percentage of genes being differentially expressed due to phenotype,
i.e. biological signal. Default is <code>0.1</code>.
If no differential expression is desired, then set <code>p.DE=0</code>.</p></td>
    </tr>
    <tr>
      <th>pLFC</th>
      <td><p>The log phenotypic fold change for DE genes. This can be:
(1) a constant, e.g. 2;
(2) a vector of values with length being number of DE genes.
If the input is a vector and the length is not the number of DE genes,
it will be sampled with replacement to generate log fold changes;
(3) a function that takes an integer n, and generates a vector of length n,
e.g. function(x) rnorm(x, mean=0, sd=1.5).
Default is <code>1</code>, i.e. a log2 fold change of 1 or fold change of 2.
Please note that the log fold change should be on <code><a href='https://rdrr.io/r/base/Log.html'>log2</a></code> scale!</p></td>
    </tr>
    <tr>
      <th>p.G</th>
      <td><p>Numeric vector indicating the proportion of replicates per group
that express the phenotypic fold change. Default is <code>1</code>, this means all show the expression difference.
For example, if <code>0.5</code> and <code>n1 = 10</code> and <code>n2 = 8</code>,
then only 5 replicates in group 1 and 4 replicates in group 2 express the phenotypic fold change.</p></td>
    </tr>
    <tr>
      <th>p.B</th>
      <td><p>Numeric vector between 0 and 1 representing the percentage of genes
being differentially expressed between batches.
Default is <code>NULL</code>, i.e. no batch effect.</p></td>
    </tr>
    <tr>
      <th>bLFC</th>
      <td><p>The log batch fold change for all genes. This can be:
(1) a constant, e.g. 2;
(2) a vector of values with length being number of all genes.
If the input is a vector and the length is not the number of total genes,
 it will be sampled with replacement to generate log fold changes;
(3) a function that takes an integer n, and generates a vector of length n,
 e.g. function(x) rnorm(x, mean=0, sd=1.5).
Note that the simulations of only two batches is implemented.
Default is <code>NULL</code>, i.e. no batch effect.
Please note that the log fold change should be on <code><a href='https://rdrr.io/r/base/Log.html'>log2</a></code> scale!</p></td>
    </tr>
    <tr>
      <th>bPattern</th>
      <td><p>Character vector for batch effect pattern if <code>p.B</code> is non-null.
Possible options include:
"uncorrelated", "orthogonal" and " correlated".
Default is <code>"uncorrelated"</code>.</p></td>
    </tr>
    <tr>
      <th>n1, n2</th>
      <td><p>Integer vectors specifying the number of biological replicates in each group.
Default values are n1=c(20,50,100) and n2=c(30,60,120).
The vectors need to have the same number of entries, i.e. length.</p></td>
    </tr>
    <tr>
      <th>Thinning</th>
      <td><p>Numeric vector specifying the downsampling.
It has to be the same length and order as the vector <code>n1</code>.
This is an implementation of <code><a href='https://rdrr.io/pkg/edgeR/man/thinCounts.html'>thinCounts</a></code>.
The vector entries should be a proportion between 0 and 1,
e.g. <code>0.75</code> means that each sample in that group
will have on average 75% sequencing depth compared
to the original sequencing depth as listed in <code><a href='estimateParam.html'>estimateParam</a></code>.
Note that no upsampling is possible, i.e. defining a proportion greater than 1.
The default is <code>NULL</code>, meaning no downsampling.</p></td>
    </tr>
    <tr>
      <th>LibSize</th>
      <td><p>Size factors representing sample-specific differences/biases
 in expected mean values of counts:
<code>"equal"</code> or <code>"given"</code>. The default is <code>"equal"</code>,
i.e. equal size factor of 1.
If the user defines it as <code>"given"</code>, the size factors are sampled from
the estimated size factors of <code><a href='estimateParam.html'>estimateParam</a></code>.</p></td>
    </tr>
    <tr>
      <th>estParamRes</th>
      <td><p>The estimated simulation parameters for genes. This can be:
(1) The output of <code><a href='estimateParam.html'>estimateParam</a></code>.
(2) A string specifying the name of precalculated estimates, see details.</p></td>
    </tr>
    <tr>
      <th>estSpikeRes</th>
      <td><p>The spike-in simulation parameters generated by <code><a href='estimateSpike.html'>estimateSpike</a></code>.
These are needed for applying spike-in-dependent normalisation methods.
Default is <code>NULL</code>, i.e. no spike-in count simulations.</p></td>
    </tr>
    <tr>
      <th>DropGenes</th>
      <td><p>By default, the estimated parameters and fit based on genes that
 were defined as expressed are used for simulations.
By setting this parameter to <code>TRUE</code>, a fraction of genes will be dropouts.
The dropout genes are defined in <code><a href='estimateParam.html'>estimateParam</a></code> using the <code>GeneFilter</code> option
and can be plotted with <code><a href='plotParam.html'>plotParam</a></code>.
Default is <code>FALSE</code>, i.e. no gene expression dropouts.</p></td>
    </tr>
    <tr>
      <th>setup.seed</th>
      <td><p>Setup seed.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical vector to indicate whether to print function information.
Default is <code>TRUE</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A complex list with the following entries:</p>
<dt>DESetup</dt><dd><p>A list of simulation options relating to the Differential Expression Setup:
number of genes simulated (ngenes); number of simulation iterations (nsims);
IDs of DE genes (DEid) which is a list (length=nsims) of vectors (length=ngenes*p.DE);
log fold change of genes (pLFC) which is a list (length=nsims) of vectors (length=ngenes);
similarly for batch effects (Bid and bLFC); list containing simulation seeds (sim.seed).</p></dd>
<dt>SimSetup</dt><dd><p>A list of simulation options relating to the Simulation Setup:
the number of samples per group (n1 and n2); simulating spike-ins (spikeIns);
Thinning parameters (Thinning, thinSpike); Library Size Factors (LibSize);
 dropout genes (DropGenes and DropRate).</p></dd>
<dt>estParamRes</dt><dd><p>A list object containing the gene simulation parameters provided.</p></dd>
<dt>estSpikeRes</dt><dd><p>A list object containing the spike-in simulation parameters provided.</p></dd>

    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Beate Vieth</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># estimate gene parameters</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"SmartSeq2_Gene_Read_Counts"</span><span class='op'>)</span>
<span class='va'>Batches</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>Batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/strsplit.html'>strsplit</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>SmartSeq2_Gene_Read_Counts</span><span class='op'>)</span>,
                                    <span class='st'>"_"</span><span class='op'>)</span>, <span class='st'>"[["</span>, <span class='fl'>1</span><span class='op'>)</span>,
                     stringsAsFactors <span class='op'>=</span> <span class='cn'>F</span>,
                     row.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>SmartSeq2_Gene_Read_Counts</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"GeneLengths_mm10"</span><span class='op'>)</span>
<span class='va'>estparam_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='estimateParam.html'>estimateParam</a></span><span class='op'>(</span>countData <span class='op'>=</span> <span class='va'>SmartSeq2_Gene_Read_Counts</span>,
                               readData <span class='op'>=</span> <span class='cn'>NULL</span>,
                               batchData <span class='op'>=</span> <span class='va'>Batches</span>,
                               spikeData <span class='op'>=</span> <span class='cn'>NULL</span>, spikeInfo <span class='op'>=</span> <span class='cn'>NULL</span>,
                               Lengths <span class='op'>=</span> <span class='va'>GeneLengths_mm10</span>, MeanFragLengths <span class='op'>=</span> <span class='cn'>NULL</span>,
                               RNAseq <span class='op'>=</span> <span class='st'>'singlecell'</span>, Protocol <span class='op'>=</span> <span class='st'>'Read'</span>,
                               Distribution <span class='op'>=</span> <span class='st'>'ZINB'</span>, Normalisation <span class='op'>=</span> <span class='st'>"scran"</span>,
                               GeneFilter <span class='op'>=</span> <span class='fl'>0.1</span>, SampleFilter <span class='op'>=</span> <span class='fl'>3</span>,
                               sigma <span class='op'>=</span> <span class='fl'>1.96</span>, NCores <span class='op'>=</span> <span class='cn'>NULL</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='co'># estimate spike parameters</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"SmartSeq2_SpikeIns_Read_Counts"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"SmartSeq2_SpikeInfo"</span><span class='op'>)</span>
<span class='va'>Batches</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>Batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/strsplit.html'>strsplit</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>SmartSeq2_SpikeIns_Read_Counts</span><span class='op'>)</span>,
                                    <span class='st'>"_"</span><span class='op'>)</span>, <span class='st'>"[["</span>, <span class='fl'>1</span><span class='op'>)</span>,
                     stringsAsFactors <span class='op'>=</span> <span class='cn'>F</span>,
                     row.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>SmartSeq2_SpikeIns_Read_Counts</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>estparam_spike</span> <span class='op'>&lt;-</span> <span class='fu'><a href='estimateSpike.html'>estimateSpike</a></span><span class='op'>(</span>spikeData <span class='op'>=</span> <span class='va'>SmartSeq2_SpikeIns_Read_Counts</span>,
                                spikeInfo <span class='op'>=</span> <span class='va'>SmartSeq2_SpikeInfo</span>,
                                MeanFragLength <span class='op'>=</span> <span class='cn'>NULL</span>,
                                batchData <span class='op'>=</span> <span class='va'>Batches</span>,
                                Normalisation <span class='op'>=</span> <span class='st'>'depth'</span><span class='op'>)</span>
<span class='co'># define log fold change</span>
<span class='va'>p.lfc</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>1</span>,<span class='fl'>1</span><span class='op'>)</span>, size<span class='op'>=</span><span class='va'>x</span>,replace<span class='op'>=</span><span class='cn'>T</span><span class='op'>)</span><span class='op'>*</span><span class='fu'><a href='https://rdrr.io/r/stats/GammaDist.html'>rgamma</a></span><span class='op'>(</span><span class='va'>x</span>, shape <span class='op'>=</span> <span class='fl'>1</span>, rate <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
<span class='co'># set up simulations</span>
<span class='va'>setupres</span> <span class='op'>&lt;-</span> <span class='fu'>Setup</span><span class='op'>(</span>ngenes <span class='op'>=</span> <span class='fl'>10000</span>, nsims <span class='op'>=</span> <span class='fl'>25</span>,
                  p.DE <span class='op'>=</span> <span class='fl'>0.1</span>, pLFC <span class='op'>=</span> <span class='va'>p.lfc</span>,
                  n1 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>20</span>,<span class='fl'>50</span>,<span class='fl'>100</span><span class='op'>)</span>, n2 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>30</span>,<span class='fl'>60</span>,<span class='fl'>120</span><span class='op'>)</span>,
                  Thinning <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>0.9</span>,<span class='fl'>0.8</span><span class='op'>)</span>, LibSize <span class='op'>=</span> <span class='st'>'given'</span>,
                  estParamRes <span class='op'>=</span> <span class='va'>estparam_gene</span>,
                  estSpikeRes <span class='op'>=</span> <span class='va'>estparam_spike</span>,
                  DropGenes <span class='op'>=</span> <span class='cn'>TRUE</span>,
                  setup.seed <span class='op'>=</span> <span class='fl'>52679</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Beate Vieth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


